### PR TITLE
Payments: add new Payments post publish panel

### PR DESCRIPTION
**Pull request by jeherve**
#### Changes proposed in this Pull Request:

﻿This adds a new editor plugin, available in the following scenarios:

- When the site is connected to WordPress.com, or on WordPress.com Simple.
- When the site uses a plan that supports the Payments feature. On WordPress.com, this will require changes to add that new Payments feature to the list mapping plans and features.
- After a post has been published.

**Notes**

* This is a first version, that @niranjan-uma-shankar can expand on later on.
* I'm not sure about the name of the new feature. It's very generic, and I worry it could conflict with upcoming Payments things. Maybe we should rename it to `payments-post-publish`? @Copons What do you think?
* This is currently only displayed for folks who have a plan. I don't know if this should display a different message to folks without a plan?
* I have not fully tested this, and more specifically, I have not tested this on pages; I believe the intent is to only show this on pages, not posts; I have not implemented this.

**Screenshot**

image

#### Jetpack product discussion

* PT: pcbrnV-3Op-p2
* Slack: p1645778718984779-slack-CJS75TX3R

#### Does this pull request change what data or activity we track or use?

* No

#### Testing instructions:

* Connect your site to WordPress.com.
* Purchase a bundled plan (like Complete)
* Go to Pages > Add New
* Write and Publish a new page.
* Notice the new plugin in the post publish panel.
* Click on the link.
* Check Tracks Live View for `jetpack_editor_payments_post_publish_click` 5 minutes later; you should see your click.
* Verify that the post publish panel _does not_ appear when publishing a new post via Posts > Add New.

### CODE REVIEW

**Comment by Copons** [comment]
> I'm not sure about the name of the new feature. It's very generic, and I worry it could conflict with upcoming Payments things. Maybe we should rename it to `payments-post-publish`?

I had the same doubt when looking at the code, as I didn't realize that the `extensions/index.json` contains both blocks and editor plugins mixed together.
Right now it's not a big deal, as we don't have a `payments` block that could conflict, but we can't exclude that we might have one in the future.
It would probably still not be a big deal anyway, as plugins can be easily renamed without breaking anything (whereas blocks would require a migration strategy).

`payments-post-publish` sounds good to me even though it might end up being limiting if we want to add other payments-related features that don't belong to the post publish panel.
At the same time, having single-features folders is generally a neat idea, with the only downside that adding new features always require some repeated scaffolding underneath.

> I have not fully tested this, and more specifically, I have not tested this on pages; I believe the intent is to only show this on pages, not posts; I have not implemented this.

Out of curiosity, is there a reason for limiting this to pages only?
I've scanned through the project thread, but couldn't find any. 🤔 
cc @sethmclellana8c

File: projects/plugins/jetpack/extensions/plugins/payments/payments.php
Comment (Copons) [inline projects/plugins/jetpack/extensions/plugins/payments/payments.php:35]: This looks a bit weird to me: we are registering an editor plugin as a block, and I'm not sure it's the right way to go — unless I'm overlooking something.

For example, the Publicize plugin too extends the publish panel (the pre-publish one, but anyway), and doesn't register as a block:

https://github.com/Automattic/jetpack/blob/master/projects/plugins/jetpack/extensions/plugins/publicize/publicize.php

Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/payments.php]: That's a good point. I'm happy to change that. Done in 9b50bb0a5c99a7ec784913481e81b844e35cb50e

**Comment by jeherve** [comment]
> I had the same doubt when looking at the code

I'm glad it's not just me. :) 

> payments-post-publish sounds good to me even though it might end up being limiting if we want to add other payments-related features that don't belong to the post publish panel.

`payments-plugin` maybe?

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (Copons) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js:43]: As an aside, do we need this to be a `Button`?
Core Gutenberg has an [`ExternalLink`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/external-link/index.js) with a few a11y niceties already built-in.

`ExternalLink` also targets `_blank` instead of `_top`, which is probably appropriate for an external link.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (Copons) [inline projects/plugins/jetpack/extensions/plugins/payments/index.js:29]: Note that `PluginPostPublishPanel` has a [`title` prop](https://github.com/WordPress/gutenberg/blob/5c9ed439ed62c6d9ce85f5d8d3aa5ceb2a6931ef/packages/edit-post/src/components/sidebar/plugin-post-publish-panel/index.js#L13) that could be helpful here.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js:43]: > It opens a page marked [private] which is not accessible to the outside world.

It is still a draft, as the contents are still a work in progress.

> As an aside, do we need this to be a Button?
> Core Gutenberg has an [ExternalLink](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/external-link/index.js?rgh-link-date=2022-02-25T18%3A04%3A08Z) with a few a11y niceties already built-in.

I had gone with `ExternalLink` at first, but ended up moving to a custom solution in 982227536a77f95dc38de47c475258505d38b656 because of the need to add a Tracks event to that link. As I mentioned in the commit message though, I'm not a big fan of that solution :)

This is the same approach we followed for the Anchor integration:
https://github.com/Automattic/jetpack/blob/d35fe2b3327a5be210c3d916df5bf3368fa742c1/projects/plugins/jetpack/extensions/blocks/anchor-fm/editor.js#L79-L90

Maybe we should move to a new component of ours, that would be core's `ExternalLink` + a Tracks implementation?

Comment (Copons) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js:43]: I haven't tried, but I _think_ `ExternalLink` supports `onClick` (or at least it should propagate all its additional props to the `<a>` inside).

So I wonder if my whole suggestion [below](https://github.com/Automattic/jetpack/pull/23118#discussion_r814977658) could be adapted to use `ExternalLink` instead of `Button`. 🤔

Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: > I think ExternalLink supports onClick

You're right, it does seem to work! That would simplify things for sure. I just pushed c2149b69288c369b1f1981dfa9b05d5fc09c95b0 to give this all a try.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: Done in c2149b69288c369b1f1981dfa9b05d5fc09c95b0

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: done in c2149b69288c369b1f1981dfa9b05d5fc09c95b0

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (Copons) [inline projects/plugins/jetpack/extensions/plugins/payments/index.js:28]: Just a random idea, good for a follow-up.
We can add customize the panel icon with the `icon` prop.

Screenshot 2022-03-01 at 12 36 00

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (Copons) [inline projects/plugins/jetpack/extensions/plugins/payments/index.js:20]: Apparently this doesn't open into `_blank`.
The hook has the logic for it already, but it's set up to only do so if `isWidgetEditor`:

https://github.com/Automattic/jetpack/blob/6e8a306f0bb3e8ce01437592c78ef4a462bde785/projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js#L12

https://github.com/Automattic/jetpack/blob/6e8a306f0bb3e8ce01437592c78ef4a462bde785/projects/plugins/jetpack/extensions/shared/use-autosave-and-redirect/index.js#L91

IMHO this is not a big enough deal to block this PR.
We can update the `useAutosaveAndRedirect` hook (or the `autosaveAndRedirect`) to also accept a `shouldOpenNewWindow` option, but it makes more sense to do it in a separate PR.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (Copons) [inline projects/plugins/jetpack/extensions/plugins/payments/index.js:31]: `--` should probably be `—`, but I'll defer to a real editorial review. 😄

**Review by Copons — APPROVED** [review]
This works well for me!
I've left a few comments, but none are blockers!

I still don't see a compelling reason to limit this to a specific post type, so as far as I'm concerned: :shipit:

File: projects/plugins/jetpack/extensions/plugins/payments/payments.php
Comment (niranjan-uma-shankar) [inline projects/plugins/jetpack/extensions/plugins/payments/payments.php:33]: @jeherve Could you please confirm if `set_availability_for_plan()` for feature name `payments` will enable this plugin for sites hosted on WP.com? Checking because [this function](https://github.com/Automattic/jetpack/blob/a27793f036312183795838b0ee8c0c03da7a9cef/projects/plugins/jetpack/class.jetpack-gutenberg.php#L1118) does not seem to map to a feature called `payments`, and only has simple-payments and recurring-payments but maybe I'm missing something.

Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/payments.php:33]: You will indeed need to update WordPress.com as well. This is why I had added this to the PR description:

> On WordPress.com, this will require changes to add that new Payments feature to the list mapping plans and features.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: Good point. 6049599541c6d8cadc99e83bc7bac7bc6f381676

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js:30]: I like the idea, thank you. Done in 8d432060ad3b1f043d66a91c0d1968c7ea171414
Screenshot 2022-03-03 at 15 52 44

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: Good point. I just committed b2cf36f67fd621d0651e9aead121b4ea5a4e3677

File: projects/plugins/jetpack/extensions/plugins/payments/payments.php
Comment (niranjan-uma-shankar) [reply projects/plugins/jetpack/extensions/plugins/payments/payments.php:33]: Ah sorry, I missed that. Thanks for clarifying.

File: projects/plugins/jetpack/extensions/plugins/payments/index.js
Comment (jeherve) [reply projects/plugins/jetpack/extensions/plugins/payments/index.js]: b0a65676b0628c9037d87d58a1cbb055a0d42d5c

**Comment by niranjan-uma-shankar** [comment]
@jeherve Is this PR good to be merged?

**Comment by jeherve** [comment]
@niranjan-uma-shankar That's up to you really, that's your PR, I was only there to keep things moving so you don't get blocked by anything. As it stands, it does get displayed regardless of the post type, it's not limited to pages. I don't know if that's a blocker for you.

**Comment by niranjan-uma-shankar** [comment]
Thanks for clarifying. 

> As it stands, it does get displayed regardless of the post type, it's not limited to pages. I don't know if that's a blocker for you.

I've requested clarification in pcbrnV-3Op-p2#comment-5660.

>  I was only there to keep things moving so you don't get blocked by anything.

Thank you! Once I hear back on pcbrnV-3Op-p2#comment-5660, I'll follow up with a request for an approval for the PR.

**Comment by niranjan-uma-shankar** [comment]
I pushed in a change to restrict the sidebar message only to page publishes, as per the feedback in pcbrnV-3Op#comment-5666.

@Copons would you be able to approve the PR if it looks good to you?

**Comment by Copons** [comment]
I took the liberty of pushing a couple of changes fixing the behaviour of the `ExternalLink`.
The core component is already built to open `_blank`, so we don't need any specific handling for that.
In particular, we don't need `useAutosaveAndRedirect` to handle the tracking as:
- We are not redirecting, but opening the link in a new window.
- We don't need to autosave because we are not navigating away (and we could also assume the page has literally just been saved when it got published 😄 ).

So, I've reverted the `useAutosaveAndRedirect` changes, and let the link do its thing with as little interference as possible.

**Review by Copons — APPROVED** [review]

**Comment by jeherve** [comment]
This can now be merged, as per pcbrnV-3Op-p2#comment-5713

**Pull request merged**
